lib/rdkafka/native_kafka.rb

```diff
@@ -51,11 +67,19 @@ def close(object_id=nil)
         @polling_thread.join
       end
 
-      # Destroy the client
-      Rdkafka::Bindings.rd_kafka_destroy_flags(
-        @inner,
-        Rdkafka::Bindings::RD_KAFKA_DESTROY_F_IMMEDIATE
-      )
+      # Run in a thread because this could be called
+      # in a trap context
+      Thread.new do
+        # Lock the mutexes for polling and method access
+        # to make sure we are not destroying during an
+        # ongoing operation
+        @method_mutex.lock
+        @poll_mutex.lock
+
+        # Destroy the client
+        Rdkafka::Bindings.rd_kafka_destroy(@inner)
+      end.join
```

Comment:
can we add ensure here to unlock? I feel bad leaving this dangling

Reply:
Implemented in a different way now